src/battle_util.c

```diff
@@ -1641,83 +1642,154 @@ static void TryToRevertMimicryAndFlags(void)
+    ENDTURN_WATER_SPORT,
+    ENDTURN_MUD_SPORT,
+    ENDTURN_WONDER_ROOM,
+    ENDTURN_MAGIC_ROOM,
+    ENDTURN_TERRAIN,
+    ENDTURN_THIRD_EVENT_BLOCK,
+    ENDTURN_EMERGENCY_EXIT_4,
+    ENDTURN_ABILITIES,
+    ENDTURN_FOURTH_EVENT_BLOCK,
+    ENDTURN_COUNT,
+};
 
-    for (gBattlerAttacker = 0; gBattlerAttacker < gBattlersCount && gAbsentBattlerFlags & gBitTable[gBattlerAttacker]; gBattlerAttacker++)
-    {
-    }
-    for (gBattlerTarget = 0; gBattlerTarget < gBattlersCount && gAbsentBattlerFlags & gBitTable[gBattlerTarget]; gBattlerTarget++)
+enum FirstEventBlock
+{
+    FIRST_EVENT_BLOCK_GMAX_MOVE_RESIDUAL,
+    FIRST_EVENT_BLOCK_SEA_OF_FIRE_DAMAGE,
+    FIRST_EVENT_BLOCK_THRASH,
+    FIRST_EVENT_BLOCK_GRASSY_TERRAIN_HEAL,
+    FIRST_EVENT_BLOCK_ABILITIES,
+    FIRST_EVENT_BLOCK_HEAL_ITEMS,
+};
+
+enum SecondEventBlock
+{
+    SECOND_EVENT_BLOCK_REFLECT,
+    SECOND_EVENT_BLOCK_LIGHT_SCREEN,
+    SECOND_EVENT_BLOCK_SAFEGUARD,
+    SECOND_EVENT_BLOCK_MIST,
+    SECOND_EVENT_BLOCK_TAILWIND,
+    SECOND_EVENT_BLOCK_LUCKY_CHANT,
+    SECOND_EVENT_BLOCK_RAINBOW,
+    SECOND_EVENT_BLOCK_SEA_OF_FIRE,
+    SECOND_EVENT_BLOCK_SWAMP,
+    SECOND_EVENT_BLOCK_AURORA_VEIL,
+};
+
+enum ThirdEventBlock
+{
+    THIRD_EVENT_BLOCK_UPROAR,
+    THIRD_EVENT_BLOCK_ABILITIES, // Bad Dreams, Ball Fetch, Harvest, Moody/ Pickup / Slow Start / Speed Boost
+    THIRD_EVENT_BLOCK_ITEMS, // Flame Orb, Sticky Barb, Toxic Orb, White Herb
+};
+
+enum FourthEventBlock
+{
+    FOURTH_EVENT_BLOCK_HUNGER_SWITCH,
+    FOURTH_EVENT_BLOCK_EJECT_PACK,
+};
+
+#define MAGIC_GUARD_CHECK \
+if (ability == ABILITY_MAGIC_GUARD) \
+{\
+    RecordAbilityBattle(battler, ability);\
+    gBattleStruct->turnEffectsBattlerId++;\
+            break;\
+}
+
+static inline u32 TryEmergencyExit(u32 battler)
+{
+    u32 ability = GetBattlerAbility(battler);
+    if (ability == ABILITY_EMERGENCY_EXIT || ability == ABILITY_WIMP_OUT)
     {
+        if (AbilityBattleEffects(ABILITYEFFECT_ENDTURN, battler, 0, 0, 0))
+            return 1;
     }
+    return 0;
+}
 
-    do
+u32 DoEndTurnEffects(void)
+{
+    u32 i = 0;
+    u32 side = 0, battler = 0;
+    u32 ability = ABILITY_NONE, holdEffect = HOLD_EFFECT_NONE;
+    u32 effect = 0;
+
+    gHitMarker |= (HITMARKER_GRUDGE | HITMARKER_IGNORE_BIDE);
+
+    while (gBattleStruct->endTurnEventsCounter < ENDTURN_COUNT)
     {
-        s32 i;
-        u32 side;
+        if (gBattleStruct->turnEffectsBattlerId == gBattlersCount || gBattleStruct->turnSideTracker == NUM_BATTLE_SIDES)
+        {
+            gBattleStruct->endTurnEventsCounter++;
+            gBattleStruct->turnEffectsBattlerId = 0;
+            gBattleStruct->turnSideTracker = 0;
+            gBattleStruct->eventBlockCounter = 0;
+        }
+
+        battler = gBattlerAttacker = gBattlerByTurnOrder[gBattleStruct->turnEffectsBattlerId];
+        ability = GetBattlerAbility(battler);
 
-        switch (gBattleStruct->turnCountersTracker)
+        switch (gBattleStruct->endTurnEventsCounter)
```

Comment:
**suggestion:** As a broader (potentially-out-of-scope) suggestion, would it make sense for each of these states to be their own function and thus be organized by a function-table? That will cost some ROM-space, but I think it will result in the underlying code being quite a bit more readable.